### DISCUSSION TITLE
several problems in hotstuff run

### DISCUSSION THREAD

**Discussion by SenjiMuramasa** [relab/hotstuff#38] [Q&A]
Hi, I recently started to learn about Hotstuff related content, and I tried to run your code.  But I ran into a few problems running the code. I will show them in the pictures. I didn't make any changes when I ran the code, I just did the following steps:
1 make
2 ./hotstuff run
But the logs seem to indicate something unusual. Do these logs represent any problems in my operation? Or are these logs normal?
Thank you very much.
![图片1](https://user-images.githubusercontent.com/96971310/153864516-cb79cee0-0902-4357-be8c-07af39a2538d.png)
![图片2](https://user-images.githubusercontent.com/96971310/153864529-7e6d0f28-0625-4ae0-8fea-7901e53c9475.png)
![图片3](https://user-images.githubusercontent.com/96971310/153864535-e126c682-159d-40bb-93bc-077f4e2bf1e2.png)
![图片4](https://user-images.githubusercontent.com/96971310/153864544-e0ecd5ff-8632-4833-a2d2-68445b95d609.png)

**Comment by johningve** [comment]
Hi!

None of these log messages indicate any serious issues. It's likely that they were all caused by messages arriving in a different order than expected. That can happen often, especially in a low-latency network.

I'll take a closer look at the sources of these log messages and consider if they should be reclassified as "debug" messages.

**Comment by SenjiMuramasa** [comment]
Thank you for your answer!
I am trying to do some experiments with your code, so I am wondering if you tested the performance of the code?
In addition, I see some content about fork in byzantine.go. Can fork be used directly? And how to use it?
I am new to hotstuff and Go, so my question may be low-level. I am very sorry to bother you.
Thank you very much.

**Comment by johningve** [comment]
> I am trying to do some experiments with your code, so I am wondering if you tested the performance of the code?

We did some comparisons with `libhotstuff` a long time ago, but I haven't looked at the performance in a while.

> In addition, I see some content about fork in byzantine.go. Can fork be used directly? And how to use it?

Uhh, I just noticed that it doesn't work on `master`. 😅I've pushed a commit to fix it!
You can enable it using the `byzantine` flag:

```text
./hotstuff run --byzantine="fork:1"
```

The example above enables the "fork" behavior for a single replica. You could combine several byzantine behaviors also:

```text
./hotstuff run --byzantine="fork:1,silence:2"
```

This would enable the "fork" behavior for one replica, and the "silence" behavior for two replicas.

**Comment by SenjiMuramasa** [comment]
Thank you very much!

I have tried using a fork attack but the code is stuck.
![fork](https://user-images.githubusercontent.com/96971310/155825727-0779bcb3-c81e-4d8c-b610-b4580543f81a.png)
So I wonder if this is a sign of a success attack? Should this trigger the timeout system to go to the next view?

And, I want to try using Batch to improve the system throughput. I changed batch-size in the run.go file, but the result was 0.
![batch1](https://user-images.githubusercontent.com/96971310/155826089-191563e0-8f7b-4b60-8d9e-bd5993a26570.png)
![batch2](https://user-images.githubusercontent.com/96971310/155826092-5ce15cfd-644a-444b-a9e0-832a999bd8c6.png)
Are there any other configuration changes that need to be made?

Again, thank you for your patient reply.

**Comment by johningve** [comment]
It looks like the default setting for `max-concurrent`, which controls the number of concurrent commands a client can issue, is too low for a fork scenario. This setting must be changed if you change the batch size. It should be at least 4 times the batch size if you are using a single client.

Going back to the fork attack, by increasing `max-concurrent` from 4 (the default) to 10, I get this result:

```text
executed: 4900, failed: 1634
```

The failed commands shows that the fork attack was successful.

Now, you can compare with the `fasthotstuff` implementation:

```
./hotstuff run --consensus="fasthotstuff" --byzantine="fork:1" --max-concurrent=10 --timeout-multiplier=1
```

(setting `--timeout-multiplier` to 1 here stops the view-synchronizer from increasing the view duration when a timeout occurs)

This gives me the result: `executed: 1510, failed: 0`, showing that the `fasthotstuff` implementation is indeed resistant to forking attacks :)

**Comment by SenjiMuramasa** [comment]
Thank you! It helps a lot. 
After increasing 'max-concurrent' from 4 to 10, the code started to run. But it seems the code can only run for a short time. I tried to continue increasing the 'max-concurrent' and it worked a little better. But it still doesn't run to the end. (I increased 'duration' to 60 seconds)
![fork](https://user-images.githubusercontent.com/96971310/155923078-f9b9edd7-abb9-4d8f-a099-3e54a67d42b1.png)
I'm using a virtual machine to run the code on Linux21. Is that the cause of the problem?

**Comment by SenjiMuramasa** [comment]
Hi, I tried to run the code many times, sometimes the code would run longer, and sometimes it would just get stuck.
According to the number of final executions and failures, the attack was successful.

These two images represent two typical cases (duration=10s, max-concurrent=10).
The first one stuck as soon as it started running.
![2](https://user-images.githubusercontent.com/96971310/156874229-4f4e1aa3-98e1-4507-9f26-7966e48d94c7.png)

The second one ran almost to the end of the duration.
![1](https://user-images.githubusercontent.com/96971310/156874275-2d716add-3540-4d52-ba7d-1eeabd01eb36.png)

Do you have this kind of problem? Or was it just me, for some other reason?

**Comment by johningve** [comment]
Sorry for the slow response. It looks to me like this is the client getting stuck due to some unacknowledged commands (or maybe unordered?). I am not quite sure why it happens. To remedy this, I have added a timeout to the client that is configurable through the `--client-timeout` flag (defaults to 500ms).

**Comment by SenjiMuramasa** [comment]
Thank you for your patience. It always takes time to deal with problems.
In my test, I found that the code kept returning from the **consensus.go** file as shown in the image below when it was stuck using a fork attack. In the **Get()** function, **Select** goes all the way to **case <- ctx.Done()**.
![1](https://user-images.githubusercontent.com/96971310/157029034-249aa1da-349b-4cf5-bd6f-234207878849.png)
![2](https://user-images.githubusercontent.com/96971310/157029053-8d0f8152-0a06-4427-bde0-67938f04ac5b.png)
My Go language level is not good enough to solve this problem. I just hope this little discovery will help.

Thanks again for your reply. You really helped me a lot.

In addition, I just started to use Github, so I did not deal with some contents that needed to be bold or marked properly. Please forgive me.

**Comment by SenjiMuramasa** [comment]
Hello, here I go again~~
I tried overwriting two blocks with a fork attack, but it seems the malicious block didn't get into the chain. Where can I modify so that the malicious blocks can be saved in the chain?
Thank you!

**Comment by johningve** [comment]
It's hard for me to tell exactly why the malicious block is not accepted without any more detail, but here is how it normally works:

1. One of the replicas creates the new block and proposes it to the other replicas. This is usually done by the `Consensus.Propose` method. However, the `ProposeRuler` interface allows the behavior of this to be changed, which is how the byzantine package works.
2. All replicas then process the block in the `OnPropose` function. This function verifies the block's quorum certificate, checks the identity of the proposer, and then checks the `VoteRule` of the consensus protocol. If all these checks pass, the command(s) stored in the block are inspected by the `Acceptor` module. If the `Acceptor` module accepts the command(s), then the block is accepted. Currently, the `Acceptor` module checks the serial numbers of commands to ensure that only new commands are accepted. This is implemented in the file `replica/cmdcache.go`
3. Depending on the consensus protocol, the block will be committed into the chain when its grandchild or great-grandchild is proposed (assuming that all blocks in between are accepted too).

PS: I'm converting this issue into a discussion, as I think that is more appropriate ;)

**Comment by SenjiMuramasa** [comment]
So, when I fork, does the number of failures in the picture below include malicious blocks?
Emmm, I mean, when I use the fork attack directly, without any modification, will the malicious block get into the chain successfully? Or will the malicious block be detected and classified into failure?

![fail](https://user-images.githubusercontent.com/96971310/158323226-23945f65-0401-4563-aedb-2672369f23ca.png)

**Reply by johningve** [reply]
I'll take another look at it later, but if I recall correctly, the answer depends on which consensus protocol is used:

- The default, `chainedhotstuff`, as well as `simplehotstuff`, do not detect the malicious blocks. Thus, the failed commands come from the legitimate blocks that were not committed due to the fork.
- `fasthotstuff` does detect the malicious blocks and rejects them. Thus, there are no failed commands when using this protocol.

**Comment by SenjiMuramasa** [comment]
Hello~
As mentioned in the readme, a tool is provided to run experiments on multiple servers over SSH. What is this tool, please? How do I use it?
Thank you~

**Reply by johningve** [reply]
It comes as part of the main `hotstuff` binary. Please take a look at the [documentation](https://github.com/relab/hotstuff/blob/master/docs/experimentation.md#running-experiments-on-remote-hosts) on how to set it up.

**Comment by SenjiMuramasa** [comment]
Hi~
I generated id_rsa file and known_hosts file by ssh, and I also modified ssh_config file.
Does the example_config.toml file needs to be modified? And are there any other files that need to be modified?
What's more, do I still run ./hotstuff run when I want to run experiments on multiple machines?
Thank you very much.

**Reply by SenjiMuramasa** [reply]
I encountered an error when I tried to run the experiment on a remote server as shown in the figure below. I tried many solutions I found on the Internet, but none of them solved the problem. Have you ever encountered this problem? If so, could you please tell me how to solve it? Thank you very much!
![problem](https://user-images.githubusercontent.com/96971310/162660677-fc1e23d5-2470-4f91-bb76-b579c64f27ab.png)

**Reply by johningve** [reply]
Maybe you can add a `HostName` field in the ssh config to specify the IP address?

For example:

```text
Host test
    HostName 127.0.0.1
    Port 1234
    User foo
    ...
```

**Comment by SenjiMuramasa** [comment]
Hi, recently I have been trying to do multi-machine experiments, but I have experienced many problems and failed to run successfully. Have you ever run multi-machine experiments? Could you please tell me the steps in detail? I would really appreciate it if you could.

**Reply by meling** [reply]
Yes, of course we have and are currently doing multi-machine experiments. We are quite busy at the moment and we may not be able to assist with these things now. Hopefully after our deadline, we will be able to provide more assistance.

However, to give you some directions for you to try... Are you able to log into your configured machines via password-less ssh login? This is a requirement for running experiments. If you don't understand what this means, please consult the internet for information.

**Comment by SenjiMuramasa** [comment]
Thank you very much. You have given me a lot of help. Please feel free to focus on your experiments. I'll list the steps I tried below, and you can help me check what might go wrong when you have a spare moment. Thank you.

1. I used `ssh-keygen -t rsa` to generate an **id_rsa** file and an **id_rsa.pub** file.
2. Then I moved the **id_rsa.pub** file to the **/home/user/.ssh** folder on the other computer by `scp`.
3. Copy **the id_rsa.pub** file to the **authorized_keys** file. In this case, two computers can be connected to each other without passwords through ssh.
4. Modify the **ssh_config** file in the **scripts** folder. The format is as follows:
```
Host 192.168.xxx.xxx hotstuff_worker_1
           User abc
           Port 22
           IdentityFile ~/.ssh/id
           UserKnownHostsFile ~/.ssh/known_hosts
```
5. Change `hosts` in the **run.go** file to the IP address of another computer.

Finally, I tried to run it and get the result shown below.
![wrong](https://user-images.githubusercontent.com/96971310/164222635-a993654c-aced-4735-b522-4eb403d06266.png)

**Reply by johningve** [reply]
Oh, that's an interesting problem! It looks like the machines you are deploying to have an older version of `glibc` than the one you are deploying from.

**Reply by johningve** [reply]
To work around this issue, I think you would have to compile hotstuff on one of the target machines, or find some way to build the binary with an older version of `glibc`. Either way, you can then tell hotstuff to upload a different binary to the remote machines by using the `--exe` flag.

**Reply by johningve** [reply]
You could also try to compile hotstuff without cgo. Try something like this

```
env CGO_ENABLED=0 go build -o ./hotstuff ./cmd/hotstuff
```

**Reply by SenjiMuramasa** [reply]
Thank you for taking time out of your busy schedule to reply. I have checked according to your suggestion, and it is true that the target computer has a lower version of the GLIBC that causes the problem. Then I tried two things.
1. I tried to compile hotstuff with `env CGO_ENABLED=0 go build -o ./hotstuff ./cmd/hotstuff`.
2. I tried to compile hotstuff on the target computer.
Both attempts yielded the same results.
![wrong](https://user-images.githubusercontent.com/96971310/164878855-a1c2f961-f16f-4485-bdec-b7b1897d0182.png)

There appears to be no error, but the application freezes after replicas are being created without any hint.
Does this indicate that there are other problems besides the GLIBC version?

**Comment by SenjiMuramasa** [comment]
Maybe there are some problems in my **ssh_config** file?
![ssh_config](https://user-images.githubusercontent.com/96971310/165296195-dcf36079-d017-44bb-94f3-1b669d47299b.png)
The path of **id_rsa** file and **known_hosts** file have been checked, and there are no mistakes.

**Comment by SenjiMuramasa** [comment]
Hi, I found the following error when I tried to run experiments with more replicas (e.g. 36 replicas on one computer).
![1](https://user-images.githubusercontent.com/96971310/166654270-bf7cca39-45d0-4998-8970-c1c5c95c9abc.png)
However, when I tried to run the experiment within ”vscode“, it worked.
![2](https://user-images.githubusercontent.com/96971310/166654489-f74916b9-b3f2-4dcf-92f6-98d11e61bc27.png)
What are the possible reasons for this? Is there a way to run this experiment directly?
